client/state/media/reducer.js

```diff
@@ -346,6 +347,136 @@ export const selectedItems = withoutPersistence( ( state = {}, action ) => {
+	 * media item, even after the item is fully saved on the server.
+	 *
+	 * This requirement means that when the server responds with a saved
+	 * media item, we need to create a mapping between the transient ID
+	 * and the actual ID of the item. This mapping allows anything still
+	 * using the transient ID to reference an already saved item to get back
+	 * the saved item rather than the trasient item.
+	 *
+	 * @param {object} state The previous state.
+	 * @param {object} action The action.
+	 * @returns {object} The next state.
+	 */
+	( state = {}, action ) => {
+		switch ( action.type ) {
+			case MEDIA_SOURCE_CHANGE: {
+				/**
+				 * Clear the media for the site.
+				 *
+				 * Dispatched when the media source changes (e.g., switching from uploaded media to
+				 * external media like Google Photos).
+				 */
+				return transformSiteTransientItems( state, action.siteId, () => ( {
+					transientItems: {},
+					transientIdsToServerIds: {},
+				} ) );
+			}
+
+			case MEDIA_ITEM_CREATE: {
+				/**
+				 * Save the transient media item.
+				 */
+				const {
+					site: { ID: siteId },
+					transientMedia,
+				} = action;
+
+				return transformSiteTransientItems(
+					state,
+					siteId,
+					( { transientItems: existingTransientItems, ...rest } ) => ( {
+						...rest,
+						transientItems: {
+							...existingTransientItems,
+							[ transientMedia.ID ]: transientMedia,
+						},
+					} )
+				);
+			}
+			case MEDIA_RECEIVE: {
+				/**
+				 * Remove the transient media item and create a mapping
+				 * between the transient ID and the saved ID.
+				 *
+				 * The `queries` reducer is responsible for saving the saved media
+				 * item into the `MediaQueryManager`.
+				 */
+				const { siteId, media: savedMedia } = action;
+
+				/**
+				 * The `transientId` property on media items is optional and when
+				 * present indicates a media item that was previously transient but
+				 * has now been persisted. Because we only care about transient media
+				 * in this reducer, if none of the received media were previously
+				 * transient, we can skip this work.
+				 */
+				const justSavedMedia = savedMedia.filter(
+					( mediaItem ) => ! isNil( mediaItem.transientId )
+				);
+
+				if ( justSavedMedia.length === 0 ) {
+					return state;
+				}
+
+				const transientItemIdsToExclude = justSavedMedia.map(
+					( mediaItem ) => mediaItem.transientId
+				);
+
+				const additionalTransientIdsToServerIds = justSavedMedia.reduce(
+					( acc, mediaItem ) => ( { ...acc, [ mediaItem.transientId ]: mediaItem.ID } ),
+					{}
+				);
+
+				return transformSiteTransientItems(
+					state,
+					siteId,
+					( { transientIdsToServerIds, transientItems: existingTransientItems } ) => ( {
+						transientIdsToServerIds: {
+							...transientIdsToServerIds,
+							...additionalTransientIdsToServerIds,
+						},
+						transientItems: omit( existingTransientItems, transientItemIdsToExclude ),
+					} )
+				);
+			}
```

Comment:
@jsnajdr @griffbrad @scinos

I've taken the approach here of having media items optionally augmented with a `transientId` property when there is a corresponding transient ID for a saved media item... however, I don't _really_ like this approach, mostly because it is a) not transparent to the consumer and b) not explicit

The only alternative I could come up with was to create a separate action, something like `promoteTransientMedia( siteId, transientId, savedMediaId )` which would do the same as this reducer case I've highlighted. That would also have the advantage of simplifying by removing the need to handle a list of items which added to the complexity of this reducer case.

Any thoughts/preferences on the approach here? Would any of y'all prefer the alternative I came up with or is there a different approach y'all can think of?

Reply:
From what I understand your suggestion makes sense to me; I don't have the full picture yet though so I would leave it to you and @jsnajdr to decide which approach makes the most sense.